src/Pirouette/SMT/Constraints.hs

```diff
@@ -1,167 +1,276 @@
 module Pirouette.SMT.Constraints where
 
+import Control.Applicative
+import Control.Monad
+import Data.Default
 import Data.Either
-import Data.List (intersperse)
-import Data.Map (Map)
+import qualified Data.Map as M
 import qualified Data.Set as S
 import Pirouette.Monad
 import Pirouette.SMT.Base
 import Pirouette.SMT.FromTerm
 import Pirouette.Term.Syntax
+import qualified Pirouette.Term.Syntax.SystemF as SystF
 import Prettyprinter hiding (Pretty (..))
 import qualified PureSMT
 
--- TODO: this module should probably be refactored somewhere;
--- I'm not entirely onboard with the 'translateData' funct as it is;
---   a) maybe we should pass the env when translating arbitrary expressions
---   b) maybe we should even keep the env in another reader layer on SolverT...
---
--- Its somthing to think about later, but for now this will do.
-
--- | Bindings from names to types (for the assign constraints)
-type Env lang = Map Name (Type lang)
-
--- | Constraints of a path during symbolic execution
--- We would like to have
--- type Constraint = Bot | And (Map Name [PirTerm]) [(Fuel, PirTerm, PirTerm))]
--- It is isomorphic to our current type, but with better access time to the variable assignements.
--- The 2 kinds of equalities are quite different,
--- one represents the complete execution of a branch, which leads to an application which cannot be further inlined
--- (Because it is a builtin or a constant),
--- whereas the other one represents an ongoing computation killed by lack of fuel.
-data AtomicConstraint lang meta
-  = Assign meta (TermMeta lang meta)
-  | VarEq meta meta
-  | NonInlinableSymbolEq (TermMeta lang meta) (TermMeta lang meta)
-  | NonInlinableSymbolNotEq (TermMeta lang meta) (TermMeta lang meta)
-  | OutOfFuelEq (TermMeta lang meta) (TermMeta lang meta)
-  | Native PureSMT.SExpr
-  deriving (Eq, Show)
+data ConstraintSet lang meta = ConstraintSet
+  { -- | Represents the set of productive symbolic variable assignments we know of.
+    -- Productive in the sense that we will always have at least one constructor
+    -- on the rhs.
+    csAssignments :: M.Map meta (TermMeta lang meta),
+    -- | Represents the symbolic variable equivalences we know of. Satisfies the invariant
+    -- that if @M.lookup v csmetaEq == Just v'@, then @v' < v@ (lexicographically)
+    csMetaEq :: M.Map meta meta,
+    -- | Any potential native constraint we might need.
+    csNative :: [PureSMT.SExpr],
+    -- | And finally, pairs of terms we've discovered /not/ to be equal
+    csRelations :: [(TermRelation, TermMeta lang meta, TermMeta lang meta)]
+  }
 
-data Constraint lang meta
-  = And [AtomicConstraint lang meta]
-  | Bot
-  deriving (Eq, Show)
+instance (Ord meta) => Semigroup (ConstraintSet lang meta) where
+  s1 <> s2 =
+    ConstraintSet
+      (csAssignments s1 <> csAssignments s2)
+      (csMetaEq s1 <> csMetaEq s2)
+      (csNative s1 <> csNative s2)
+      (csRelations s1 <> csRelations s2)
```

Comment:
This will take the left-biased union of assignments and equalities, erasing keys of `s2` that are already present in `s1`. Isn't that a bit dangerous? Or does it actually not pose any problem ever? Also, I am not convinced that it conserves the invariants of the map of equalities. (It might even lose some equalities?)

I'm in the process of replaccing `csAssignments` and `csMetaEq` by a union-find implementation, so that should take care of the latter. For the former, wouldn't it make more sense to require `Semigroup meta` and to use `unionWith (<>)`? Will we even have `Semigroup meta`?